crates/pep508-rs/src/marker.rs

```diff
@@ -899,195 +899,437 @@ impl<'a> TryFrom<MarkerEnvironmentBuilder<'a>> for MarkerEnvironment {
     }
 }
 
-/// Represents one clause such as `python_version > "3.8"` in the form
-/// ```text
-/// <a name from the PEP508 list | a string> <an operator> <a name from the PEP508 list | a string>
-/// ```
+/// Represents one clause such as `python_version > "3.8"`.
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub struct MarkerExpression {
-    /// A name from the PEP508 list or a string
-    pub l_value: MarkerValue,
-    /// an operator, such as `>=` or `not in`
-    pub operator: MarkerOperator,
-    /// A name from the PEP508 list or a string
-    pub r_value: MarkerValue,
+#[allow(missing_docs)]
+pub enum MarkerExpression {
+    /// <version key> <version op> <quoted PEP 440 version>
+    Version {
+        key: MarkerValueVersion,
+        specifier: VersionSpecifier,
+    },
+    /// <quoted PEP 440 version> <version op> <version key>
+    VersionInverted {
+        /// No star allowed here, `'3.*' == python_version` is not a valid PEP 440 comparison.
+        version: Version,
+        operator: pep440_rs::Operator,
+        key: MarkerValueVersion,
+    },
+    /// An string marker comparison, e.g. `sys_platform == '...'`.
+    String {
+        key: MarkerValueString,
+        operator: MarkerOperator,
+        value: String,
+    },
+    /// An string marker comparison, e.g. `'...' == sys_platform`.
+    StringInverted {
+        value: String,
+        operator: MarkerOperator,
+        key: MarkerValueString,
+    },
+    /// `extra <extra op> '...'` or `'...' <extra op> extra`
+    Extra {
```

Comment:
Do we need an inverted variant here? Without it the parser becomes lossy and won't preserve the ordering after parsing (which might be useful for reconstructing the input using `Display`), but for our use case it's unnecessary.

Reply:
Fine by me